src/main/java/com/github/firmwehr/gentle/firm/optimization/GlobalValueNumbering.java

```diff
@@ -0,0 +1,374 @@
+				case iro_Store -> {
+					var n0 = (Store) node;
+					var n1 = (Store) thatNode;
+					if (!n0.getType().equals(n1.getType())) {
+						return false;
+					}
+				}
+				case iro_Switch -> throw new InternalCompilerException("switch node is not supported");
+				case iro_Jmp, iro_Proj, iro_Start, iro_Cond -> {
+					/* some nodes should never be equal, unless firm itself considers them equal, but this way already
+					 * check a few lines above. if we reached this place, they are simply considered not equal
+					 */
+					return false;
+				}
+				default -> {/*no op*/}
+			}
+
+			return true;
+		}
+
+		@Override
+		public int hashCode() {
+			// nodes are considered equal if they themself are equally configured and share the same preds
+			// this hash ignores node configuration, but that's okay
+			int hash = node.getClass().hashCode();
+			for (Node pred : node.getPreds()) {
+				hash ^= pred.ptr.hashCode();
+			}
+
+			return hash;
+		}
+
+
+	}
+
+	/**
+	 * This class uses the Firm backend to perform a depth first search in the dominator tree. During the search, we
+	 * keep a list of available expressions (nodes) and each time we enter a new block, we check if we have duplicated
+	 * expressions. On exit, we remove the available expressions from the available expression list. This ensures that
+	 * the available expression list will only contain such expressions that are currently available to the current
+	 * block.
+	 */
+	private static class GlobalValueNumberingWalker {
+
+		// firm can't enumerate all nodes in block without shitting itself
+		private final ArrayListMultimap<Block, Node> blocks = ArrayListMultimap.create();
+		private final HashMap<NodeHashKey, Node> availableExpressions = new HashMap<>();
+		private final Set<Node> emergencyIntelligenceIncinerator = new HashSet<>();
+
+		private final Graph graph;
+		private final Set<Block> openBlocks = new HashSet<>();
+
+		private boolean hasModifiedGraph;
+
+		public GlobalValueNumberingWalker(Graph graph) {
+			this.graph = graph;
+		}
+
+		public boolean hasModifiedGraph() {
+			return hasModifiedGraph;
+		}
+
+		public void walkTheWalk() {
+			graph.walkTopological(new Default() {
+				@Override
+				public void defaultVisit(Node n) {
+
+					Block block = (Block) n.getBlock();
+					if (block != null) {
+						blocks.put(block, n);
+					}
+				}
+			});
+
+			GentleBindings.walkDominatorTree(graph, this::onEnter, this::onExit);
+
+			// remove killed nodes (don't worry, they aren't sentient)
+			for (var node : emergencyIntelligenceIncinerator) {
+				LOGGER.debug("killing %s", node);
+				Graph.killNode(node);
+			}
+			firm.bindings.binding_irgopt.remove_bads(graph.ptr);
+		}
+
+		/**
+		 * Called when DFS enters block.
+		 *
+		 * @param block The block we just entered.
+		 */
+		private void onEnter(Block block) {
+			openBlocks.add(block);
+
+			var nodes = blocks.get(block);
+
+			// stabalize local block
```

Comment:
```suggestion
			// stabilize local block
```
(we don't stab Alize here)

Reply:
Too late, she already dieded